### ISSUE TITLE
Error initializing database

### ISSUE THREAD

**Issue by dylanmcreynolds** [bluesky/tiled#616]
Running the va109 docker image, I exec into the container and try to initialize the database using

```
$ tiled admin initialize-database postgresql+asyncpg://user:pass@postgres/tiled
```

It looks like it's talking to the database just fine, but then blows up with some sort of sqlalchemy error.

```

Database postgresql+asyncpg://tiled:***@postgres/tiled is new. Creating tables and marking revision c7bd2573716d.
Traceback (most recent call last):

  File "/opt/venv/lib/python3.11/site-packages/tiled/commandline/_admin.py", line 35, in do_setup
    await check_database(engine, REQUIRED_REVISION, ALL_REVISIONS)

  File "/opt/venv/lib/python3.11/site-packages/tiled/alembic_utils.py", line 139, in check_database
    raise UninitializedDatabase(

tiled.alembic_utils.UninitializedDatabase: The database postgresql+asyncpg://tiled:***@postgres/tiled has no revision stamp. It may be empty. It can be initialized with `initialize_database(engine)`.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 528, in _prepare_and_execute
    prepared_stmt, attributes = await adapt_connection._prepare(
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 779, in _prepare
    prepared_stmt = await self._connection.prepare(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/asyncpg/connection.py", line 565, in prepare
    return await self._prepare(
           ^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/asyncpg/connection.py", line 583, in _prepare
    stmt = await self._get_statement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/asyncpg/connection.py", line 397, in _get_statement
    statement = await self._protocol.prepare(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

asyncpg.exceptions.UndefinedTableError: relation "roles" does not exist


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2112, in _exec_insertmany_context
    dialect.do_execute(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 922, in do_execute
    cursor.execute(statement, parameters)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 586, in execute
    self._adapt_connection.await_(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 125, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 185, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 564, in _prepare_and_execute
    self._handle_exception(error)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 515, in _handle_exception
    self._adapt_connection._handle_exception(error)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 803, in _handle_exception
    raise translated_error from error

sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.ProgrammingError: <class 'asyncpg.exceptions.UndefinedTableError'>: relation "roles" does not exist


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/opt/venv/bin/tiled", line 8, in <module>
    sys.exit(main())
             ^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/tiled/commandline/_admin.py", line 49, in initialize_database
    asyncio.run(do_setup())

  File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/tiled/commandline/_admin.py", line 42, in do_setup
    await initialize_database(engine)

  File "/opt/venv/lib/python3.11/site-packages/tiled/authn_database/core.py", line 70, in initialize_database
    await create_default_roles(db)

  File "/opt/venv/lib/python3.11/site-packages/tiled/authn_database/core.py", line 57, in create_default_roles
    await db.commit()

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/session.py", line 1004, in commit
    await greenlet_spawn(self.sync_session.commit)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 192, in greenlet_spawn
    result = context.switch(value)
             ^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1967, in commit
    trans.commit(_to_root=True)

  File "<string>", line 2, in commit

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1256, in commit
    self._prepare_impl()

  File "<string>", line 2, in _prepare_impl

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1231, in _prepare_impl
    self.session.flush()

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4310, in flush
    self._flush(objects)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4445, in _flush
    with util.safe_reraise():

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4406, in _flush
    flush_context.execute()

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
    util.preloaded.orm_persistence.save_obj(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 93, in save_obj
    _emit_insert_statements(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 1136, in _emit_insert_statements
    result = connection.execute(
             ^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1416, in execute
    return meth(
           ^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 516, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1639, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1843, in _execute_context
    return self._exec_insertmany_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2120, in _exec_insertmany_context
    self._handle_dbapi_exception(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2343, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2112, in _exec_insertmany_context
    dialect.do_execute(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 922, in do_execute
    cursor.execute(statement, parameters)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 586, in execute
    self._adapt_connection.await_(

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 125, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 185, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 564, in _prepare_and_execute
    self._handle_exception(error)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 515, in _handle_exception
    self._adapt_connection._handle_exception(error)

  File "/opt/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 803, in _handle_exception
    raise translated_error from error

sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedTableError'>: relation "roles" does not exist
[SQL: INSERT INTO roles (name, description, scopes, time_updated) SELECT p0::VARCHAR, p1::VARCHAR, p2::VARCHAR, p3::TIMESTAMP WITHOUT TIME ZONE FROM (VALUES ($1::VARCHAR, $2::VARCHAR, $3::VARCHAR, $4::TIMESTAMP WITHOUT TIME ZONE, 0), ($5::VARCHAR, $6::VARCHAR, $7::VARCHAR, $8::TIMESTAMP WITHOUT TIME ZONE, 1)) AS imp_sen(p0, p1, p2, p3, sen_counter) ORDER BY sen_counter RETURNING roles.id, roles.time_created, roles.id AS id__1]
[parameters: ('user', 'Default Role for users.', '["read:metadata", "read:data", "create", "write:metadata", "write:data", "apikeys"]', None, 'admin', 'Role with elevated privileges.', '["read:metadata", "read:data", "create", "write:metadata", "write:data", "admin:apikeys", "read:principals", "metrics"]', None)]
(Background on this error at: https://sqlalche.me/e/20/f405)

```

As a side note, is this documentation old? When I tried sending either flag `--if-not-exists, --no-if-not-exists` I got an error that it's not supported:

```
$ tiled admin initialize-database --if-not-exists  postgresql+asyncpg://user:pass@postgres/tiled
Usage: tiled admin initialize-database [OPTIONS] DATABASE_URI
Try 'tiled admin initialize-database --help' for help.

Error: No such option: --if-not-exists

```

**Comment by dylanmcreynolds** [comment]
Yes, yes it does. Thanks!

**Comment by kleinhenz** [comment]
I just ran into this with latest tiled and postgres as well. Specifically getting the 'relation "roles" does not exist' error. I think this is still a problem since the catalog database is not the same as the authn database unless I'm missing something. It seems to work correctly with sqlite.

**Comment by kleinhenz** [comment]
This patch seems to fix the issue. It seems like for postgres the create_all transaction doesn't get automatically committed.

```
diff --git a/tiled/authn_database/core.py b/tiled/authn_database/core.py
index 1702b26..9cfecab 100644
--- a/tiled/authn_database/core.py
+++ b/tiled/authn_database/core.py
@@ -66,6 +66,7 @@ async def initialize_database(engine):
     async with engine.connect() as conn:
         # Create all tables.
         await conn.run_sync(Base.metadata.create_all)
+        await conn.commit()

         # Initialize Roles table.
         async with AsyncSession(engine) as db:
```

**Comment by danielballan** [comment]
@kleinhenz! Nice to see you around here!

I opened #703 with this patch. We test the _catalog_ database against both SQLite and PG. This points to a need to do the same for the auth database.

**Comment by kleinhenz** [comment]
closed by #703
